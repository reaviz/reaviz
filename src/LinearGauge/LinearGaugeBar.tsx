import React, { FC } from 'react';
import { Bar, BarProps } from '../BarChart';
import { ChartTooltip, TooltipTemplate } from '@/common/Tooltip';

export type LinearGaugeBarProps = BarProps;

export const LinearGaugeBar: FC<Partial<LinearGaugeBarProps>> = (props) => (
  <Bar
    tooltip={
      <ChartTooltip
        placement="top"
        content={(data) => (
          <TooltipTemplate value={{ y: data.value, x: data.y }} />
        )}
      />
    }
    {...props}
  />
);
