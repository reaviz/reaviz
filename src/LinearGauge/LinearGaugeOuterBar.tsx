import React, { FC } from 'react';

export interface LinearGaugeOuterBarProps extends SVGRect {
  /**
   * Height of the chart. Set by `LinearGauge` component.
   */
  height: number;

  /**
   * Width of the chart. Set by `LinearGauge` component.
   */
  width: number;

  /**
   * Fill for the bar element.
   */
  fill: string;
}

export const LinearGaugeOuterBar: FC<Partial<LinearGaugeOuterBarProps>> = ({
  height,
  width,
  fill = '#484848',
  ...rest
}) => (
  <rect
    {...rest}
    fill={fill}
    width={Math.max(width!, 0)}
    height={Math.max(height!, 0)}
  />
);
