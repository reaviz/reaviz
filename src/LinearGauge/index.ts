export * from './LinearGauge';
export * from './LinearGaugeBar';
export * from './LinearGaugeOuterBar';
export * from './LinearGaugeSeries';
