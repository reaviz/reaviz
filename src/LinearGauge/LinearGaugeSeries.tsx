import React, { Fragment, ReactElement, FC } from 'react';
import { BarSeries, BarSeriesProps } from '@/BarChart';
import { LinearGaugeBar, LinearGaugeBarProps } from './LinearGaugeBar';
import {
  LinearGaugeOuterBar,
  LinearGaugeOuterBarProps
} from './LinearGaugeOuterBar';
import { CloneElement } from 'reablocks';

export interface LinearGaugeSeriesProps extends Omit<BarSeriesProps, 'bar'> {
  /**
   * Height of the chart. Set by `LinearGauge` component.
   */
  height: number;

  /**
   * Width of the chart. Set by `LinearGauge` component.
   */
  width: number;

  /**
   * Whether its a multi-series or not. Set by `LinearGauge` component.
   */
  isMultiSeries: boolean;

  /**
   * Bar element. Only applicable for single series.
   */
  bar: ReactElement<LinearGaugeBarProps, typeof LinearGaugeBar>;

  /**
   * Outer bar element. Only applicable for single series.
   */
  outerBar: ReactElement<LinearGaugeOuterBarProps, typeof LinearGaugeOuterBar>;
}

export const LinearGaugeSeries: FC<Partial<LinearGaugeSeriesProps>> = ({
  height,
  width,
  bar = <LinearGaugeBar />,
  outerBar = <LinearGaugeOuterBar />,
  isMultiSeries,
  ...rest
}) => (
  <Fragment>
    {!isMultiSeries && outerBar && (
      <CloneElement<LinearGaugeOuterBarProps>
        element={outerBar}
        height={height}
        width={width}
      />
    )}
    <BarSeries
      {...rest}
      layout="horizontal"
      tooltip={null}
      bar={<CloneElement<LinearGaugeOuterBarProps> element={bar} />}
    />
  </Fragment>
);
