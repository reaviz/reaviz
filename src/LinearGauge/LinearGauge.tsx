import React, { FC, Fragment, ReactElement, useCallback, useMemo } from 'react';
import {
  ChartProps,
  ChartContainer,
  ChartContainerChildProps
} from '@/common/containers';
import {
  ChartShallowDataShape,
  buildBarStackData,
  ChartInternalNestedDataShape,
  buildShallowChartData,
  ChartNestedDataShape
} from '@/common/data';
import { getXScale, getYScale } from '@/common/scales';
import { CloneElement } from 'reablocks';
import { LinearGaugeSeries, LinearGaugeSeriesProps } from './LinearGaugeSeries';

export interface LinearGaugeProps extends ChartProps {
  /**
   * Data the chart will receive to render.
   */
  data: ChartShallowDataShape | ChartShallowDataShape[];

  /**
   * The series component that renders the bar components.
   */
  series: ReactElement<LinearGaugeSeriesProps, typeof LinearGaugeSeries>;

  /**
   * Min value to scale on. Only applicable for single-series.
   */
  minValue: number;

  /**
   * Max value to scale on. Only applicable for single-series.
   */
  maxValue: number;
}

export const LinearGauge: FC<Partial<LinearGaugeProps>> = ({
  id,
  width,
  height,
  margins,
  className,
  containerClassName,
  series = <LinearGaugeSeries />,
  data,
  minValue = 0,
  maxValue = 100
}) => {
  const transformedData = useMemo(() => {
    if (Array.isArray(data)) {
      return buildBarStackData(
        [
          {
            key: 'default',
            data
          }
        ] as ChartNestedDataShape[],
        'expand',
        'horizontal'
      );
    } else {
      return buildShallowChartData([data], 'horizontal');
    }
  }, [data]);

  const getScales = useCallback(
    (
      isMultiSeries: boolean,
      data: ChartInternalNestedDataShape[],
      width: number,
      height: number,
      minValue: number,
      maxValue: number
    ) => {
      const domain = !isMultiSeries ? [minValue, maxValue] : undefined;

      const keyScale = getXScale({
        width,
        type: 'value',
        data,
        domain,
        isMultiSeries
      });

      const valueScale = getYScale({
        type: 'category',
        height,
        data,
        isMultiSeries
      });

      return {
        keyScale,
        valueScale
      };
    },
    []
  );

  const renderChart = useCallback(
    ({ chartHeight, chartWidth, id, chartSized }: ChartContainerChildProps) => {
      const isMultiSeries = Array.isArray(data);
      const type = isMultiSeries ? 'stackedNormalized' : 'standard';
      const { keyScale, valueScale } = getScales(
        isMultiSeries,
        transformedData as ChartInternalNestedDataShape[],
        chartWidth,
        chartHeight,
        minValue,
        maxValue
      );

      return (
        <Fragment>
          {chartSized && (
            <CloneElement<LinearGaugeSeriesProps>
              element={series}
              id={`linear-gauge-series-${id}`}
              data={transformedData}
              isCategorical={true}
              xScale={keyScale}
              yScale={valueScale}
              type={type}
              height={chartHeight}
              width={chartWidth}
              isMultiSeries={isMultiSeries}
            />
          )}
        </Fragment>
      );
    },
    [data, getScales, maxValue, minValue, series, transformedData]
  );

  return (
    <ChartContainer
      id={id}
      width={width}
      height={height}
      margins={margins}
      className={className}
      containerClassName={containerClassName}
    >
      {renderChart}
    </ChartContainer>
  );
};
