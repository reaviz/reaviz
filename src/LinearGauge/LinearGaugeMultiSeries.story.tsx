import React from 'react';
import { LinearGauge } from './LinearGauge';
import { DiscreteLegend, DiscreteLegendEntry } from '@/common/legends';
import { mitreData } from 'reaviz-data-utils';
import { LinearGaugeSeries } from './LinearGaugeSeries';
import chroma from 'chroma-js';
import { LinearGaugeBar } from './LinearGaugeBar';
import { LinearGaugeOuterBar } from './LinearGaugeOuterBar';

const colorScheme = chroma
  .scale(['#FF9C02', '#870000'])
  .colors(mitreData.length);

// Credit: https://icons8.com/icons/set/access
const Icon = ({ fill }) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 30 30"
    width="60"
    height="60"
    fill={fill}
  >
    <path d="M 15 3 C 10.398 3 5.6113281 4.0371094 5.6113281 4.0371094 L 5.6074219 4.0410156 A 2 2 0 0 0 4 6 L 4 12 C 4 22.398 14.021484 26.744141 14.021484 26.744141 A 2 2 0 0 0 15 27 A 2 2 0 0 0 15.974609 26.744141 L 15.978516 26.744141 C 15.979516 26.744141 26 22.398 26 12 L 26 6 A 2 2 0 0 0 24.388672 4.0371094 C 24.388672 4.0371094 19.602 3 15 3 z M 15 7 C 18.309 7 21 9.691 21 13 C 21 16.309 18.309 19 15 19 C 11.691 19 9 16.309 9 13 C 9 9.691 11.691 7 15 7 z M 15 9 C 14.26 9 13.575469 9.2144062 12.980469 9.5664062 L 18.433594 15.019531 C 18.785594 14.424531 19 13.74 19 13 C 19 10.794 17.206 9 15 9 z M 11.566406 10.980469 C 11.214406 11.575469 11 12.26 11 13 C 11 15.206 12.794 17 15 17 C 15.74 17 16.424531 16.785594 17.019531 16.433594 L 11.566406 10.980469 z" />
  </svg>
);

export default {
  tags: ['snapshot'],
  title: 'Charts/Gauge/Linear/Multi-Series',
  component: LinearGauge,
  subcomponents: {
    LinearGaugeSeries,
    LinearGauge,
    LinearGaugeBar,
    LinearGaugeOuterBar
  }
};

export const _Simple = () => (
  <div style={{ width: '465px', textAlign: 'center' }}>
    <h2 style={{ color: 'white', margin: 0 }}>MIRTE Attack</h2>
    <LinearGauge
      id="simple"
      height={30}
      data={mitreData}
      series={<LinearGaugeSeries colorScheme={colorScheme} />}
    />
    <DiscreteLegend
      orientation="horizontal"
      entries={mitreData.map((v, i) => (
        <DiscreteLegendEntry
          key={v.key}
          style={{ padding: '0 3px' }}
          symbol={<Icon fill={colorScheme[i]} />}
          label={`${v.data}`}
          color={colorScheme[i]}
          orientation="horizontal"
        />
      ))}
    />
  </div>
);
