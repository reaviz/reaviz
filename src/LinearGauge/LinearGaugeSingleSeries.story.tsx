import React, { useState } from 'react';
import { LinearGauge } from './LinearGauge';
import { LinearGaugeSeries } from './LinearGaugeSeries';
import { LinearGaugeBar } from './LinearGaugeBar';
import { LinearGaugeOuterBar } from './LinearGaugeOuterBar';

export default {
  tags: ['snapshot'],
  title: 'Charts/Gauge/Linear/Single-Series',
  component: LinearGauge,
  subcomponents: {
    LinearGaugeSeries,
    LinearGauge,
    LinearGaugeBar,
    LinearGaugeOuterBar
  }
};

export const Simple = () => (
  <div style={{ textAlign: 'center' }}>
    <h2 style={{ color: 'white', margin: 0 }}>Risk Score</h2>
    <LinearGauge
      id="simple"
      height={30}
      width={300}
      data={{ key: 'Risk Score', data: 80 }}
    />
  </div>
);

export const AutoSize = () => {
  const [width, setWidth] = useState(300);
  return (
    <div style={{ textAlign: 'center' }}>
      <div style={{ width, padding: 5, border: 'solid 1px red' }}>
        <LinearGauge
          id="auto-size"
          height={20}
          data={{ key: 'Risk Score', data: 80 }}
        />
      </div>
      <br />
      <button type="button" onClick={() => setWidth(width === 300 ? 150 : 300)}>
        Update Size
      </button>
    </div>
  );
};

export const NonZeroStart = () => (
  <div style={{ textAlign: 'center' }}>
    <LinearGauge
      id="non-zero-start"
      height={30}
      width={300}
      data={{ key: 'Risk Score', data: [15, 80] }}
    />
  </div>
);

NonZeroStart.story = {
  name: 'Non-Zero Start'
};

export const MultipleGauges = () => (
  <div style={{ textAlign: 'center' }}>
    <LinearGauge
      id="multiple-gauges-1"
      height={15}
      width={300}
      data={{ key: 'Step 1', data: [0, 15] }}
    />
    <br />
    <LinearGauge
      id="multiple-gauges-2"
      height={15}
      width={300}
      data={{ key: 'Step 2', data: [15, 50] }}
    />
    <br />
    <LinearGauge
      id="multiple-gauges-3"
      height={15}
      width={300}
      data={{ key: 'Step 3', data: [50, 55] }}
    />
    <br />
    <LinearGauge
      id="multiple-gauges-4"
      height={15}
      width={300}
      data={{ key: 'Step 4', data: [55, 100] }}
    />
  </div>
);
