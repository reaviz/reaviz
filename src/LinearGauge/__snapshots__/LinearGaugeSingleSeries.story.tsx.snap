// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charts/Gauge/Linear/Single-Series AutoSize smoke-test 1`] = `
<div style="text-align: center;">
  <div style="width: 300px; padding: 5px; border: 1px solid red;">
    <div class="_container_1u3dt_1"
         height="20"
         style="height: 20px; width: 100%;"
    >
      <svg width="288"
           height="20"
           class="_svg_1u3dt_30"
           tabindex="0"
      >
        <g transform="translate(0, 0)">
          <rect fill="#484848"
                width="288"
                height="20"
          >
          </rect>
          <g>
            <rect class
                  mask
                  rx="0"
                  ry="0"
                  tabindex="0"
                  aria-label="80: Risk Score"
                  role="graphics-document"
                  height="20px"
                  width="230px"
                  fill="url(#gradient-linear-gauge-series-auto-size-bar-undefined-0)"
                  x="0"
                  y="0"
                  style="cursor: auto;"
            >
            </rect>
          </g>
          <linearGradient spreadmethod="pad"
                          id="gradient-linear-gauge-series-auto-size-bar-undefined-0"
                          y1="0%"
                          y2="0%"
                          x1="0%"
                          x2="100%"
          >
            <stop offset="0%"
                  stop-opacity="0.3"
                  stop-color="#2d60e8"
            >
            </stop>
            <stop offset="80%"
                  stop-opacity="1"
                  stop-color="#2d60e8"
            >
            </stop>
          </linearGradient>
        </g>
      </svg>
    </div>
  </div>
  <br>
  <button type="button">
    Update Size
  </button>
</div>
`;

exports[`Charts/Gauge/Linear/Single-Series MultipleGauges smoke-test 1`] = `
<div style="text-align: center;">
  <div class="_container_1u3dt_1"
       width="300"
       height="15"
       style="height: 15px; width: 300px;"
  >
    <svg width="300"
         height="15"
         class="_svg_1u3dt_30"
         tabindex="0"
    >
      <g transform="translate(0, 0)">
        <rect fill="#484848"
              width="300"
              height="15"
        >
        </rect>
        <g>
          <rect class
                mask
                rx="0"
                ry="0"
                tabindex="0"
                aria-label="15: Step 1"
                role="graphics-document"
                height="15px"
                width="45px"
                fill="url(#gradient-linear-gauge-series-multiple-gauges-1-bar-undefined-0)"
                x="0"
                y="0"
                style="cursor: auto;"
          >
          </rect>
        </g>
        <linearGradient spreadmethod="pad"
                        id="gradient-linear-gauge-series-multiple-gauges-1-bar-undefined-0"
                        y1="0%"
                        y2="0%"
                        x1="0%"
                        x2="100%"
        >
          <stop offset="0%"
                stop-opacity="0.3"
                stop-color="#2d60e8"
          >
          </stop>
          <stop offset="80%"
                stop-opacity="1"
                stop-color="#2d60e8"
          >
          </stop>
        </linearGradient>
      </g>
    </svg>
  </div>
  <br>
  <div class="_container_1u3dt_1"
       width="300"
       height="15"
       style="height: 15px; width: 300px;"
  >
    <svg width="300"
         height="15"
         class="_svg_1u3dt_30"
         tabindex="0"
    >
      <g transform="translate(0, 0)">
        <rect fill="#484848"
              width="300"
              height="15"
        >
        </rect>
        <g>
          <rect class
                mask
                rx="0"
                ry="0"
                tabindex="0"
                aria-label="50: Step 2"
                role="graphics-document"
                height="15px"
                width="105px"
                fill="url(#gradient-linear-gauge-series-multiple-gauges-2-bar-undefined-0)"
                x="45"
                y="0"
                style="cursor: auto;"
          >
          </rect>
        </g>
        <linearGradient spreadmethod="pad"
                        id="gradient-linear-gauge-series-multiple-gauges-2-bar-undefined-0"
                        y1="0%"
                        y2="0%"
                        x1="0%"
                        x2="100%"
        >
          <stop offset="0%"
                stop-opacity="0.3"
                stop-color="#2d60e8"
          >
          </stop>
          <stop offset="80%"
                stop-opacity="1"
                stop-color="#2d60e8"
          >
          </stop>
        </linearGradient>
      </g>
    </svg>
  </div>
  <br>
  <div class="_container_1u3dt_1"
       width="300"
       height="15"
       style="height: 15px; width: 300px;"
  >
    <svg width="300"
         height="15"
         class="_svg_1u3dt_30"
         tabindex="0"
    >
      <g transform="translate(0, 0)">
        <rect fill="#484848"
              width="300"
              height="15"
        >
        </rect>
        <g>
          <rect class
                mask
                rx="0"
                ry="0"
                tabindex="0"
                aria-label="55: Step 3"
                role="graphics-document"
                height="15px"
                width="15px"
                fill="url(#gradient-linear-gauge-series-multiple-gauges-3-bar-undefined-0)"
                x="150"
                y="0"
                style="cursor: auto;"
          >
          </rect>
        </g>
        <linearGradient spreadmethod="pad"
                        id="gradient-linear-gauge-series-multiple-gauges-3-bar-undefined-0"
                        y1="0%"
                        y2="0%"
                        x1="0%"
                        x2="100%"
        >
          <stop offset="0%"
                stop-opacity="0.3"
                stop-color="#2d60e8"
          >
          </stop>
          <stop offset="80%"
                stop-opacity="1"
                stop-color="#2d60e8"
          >
          </stop>
        </linearGradient>
      </g>
    </svg>
  </div>
  <br>
  <div class="_container_1u3dt_1"
       width="300"
       height="15"
       style="height: 15px; width: 300px;"
  >
    <svg width="300"
         height="15"
         class="_svg_1u3dt_30"
         tabindex="0"
    >
      <g transform="translate(0, 0)">
        <rect fill="#484848"
              width="300"
              height="15"
        >
        </rect>
        <g>
          <rect class
                mask
                rx="0"
                ry="0"
                tabindex="0"
                aria-label="100: Step 4"
                role="graphics-document"
                height="15px"
                width="135px"
                fill="url(#gradient-linear-gauge-series-multiple-gauges-4-bar-undefined-0)"
                x="165"
                y="0"
                style="cursor: auto;"
          >
          </rect>
        </g>
        <linearGradient spreadmethod="pad"
                        id="gradient-linear-gauge-series-multiple-gauges-4-bar-undefined-0"
                        y1="0%"
                        y2="0%"
                        x1="0%"
                        x2="100%"
        >
          <stop offset="0%"
                stop-opacity="0.3"
                stop-color="#2d60e8"
          >
          </stop>
          <stop offset="80%"
                stop-opacity="1"
                stop-color="#2d60e8"
          >
          </stop>
        </linearGradient>
      </g>
    </svg>
  </div>
</div>
`;

exports[`Charts/Gauge/Linear/Single-Series NonZeroStart smoke-test 1`] = `
<div style="text-align: center;">
  <div class="_container_1u3dt_1"
       width="300"
       height="30"
       style="height: 30px; width: 300px;"
  >
    <svg width="300"
         height="30"
         class="_svg_1u3dt_30"
         tabindex="0"
    >
      <g transform="translate(0, 0)">
        <rect fill="#484848"
              width="300"
              height="30"
        >
        </rect>
        <g>
          <rect class
                mask
                rx="0"
                ry="0"
                tabindex="0"
                aria-label="80: Risk Score"
                role="graphics-document"
                height="30px"
                width="195px"
                fill="url(#gradient-linear-gauge-series-non-zero-start-bar-undefined-0)"
                x="45"
                y="0"
                style="cursor: auto;"
          >
          </rect>
        </g>
        <linearGradient spreadmethod="pad"
                        id="gradient-linear-gauge-series-non-zero-start-bar-undefined-0"
                        y1="0%"
                        y2="0%"
                        x1="0%"
                        x2="100%"
        >
          <stop offset="0%"
                stop-opacity="0.3"
                stop-color="#2d60e8"
          >
          </stop>
          <stop offset="80%"
                stop-opacity="1"
                stop-color="#2d60e8"
          >
          </stop>
        </linearGradient>
      </g>
    </svg>
  </div>
</div>
`;

exports[`Charts/Gauge/Linear/Single-Series Simple smoke-test 1`] = `
<div style="text-align: center;">
  <h2 style="color: white; margin: 0px;">
    Risk Score
  </h2>
  <div class="_container_1u3dt_1"
       width="300"
       height="30"
       style="height: 30px; width: 300px;"
  >
    <svg width="300"
         height="30"
         class="_svg_1u3dt_30"
         tabindex="0"
    >
      <g transform="translate(0, 0)">
        <rect fill="#484848"
              width="300"
              height="30"
        >
        </rect>
        <g>
          <rect class
                mask
                rx="0"
                ry="0"
                tabindex="0"
                aria-label="80: Risk Score"
                role="graphics-document"
                height="30px"
                width="240px"
                fill="url(#gradient-linear-gauge-series-simple-bar-undefined-0)"
                x="0"
                y="0"
                style="cursor: auto;"
          >
          </rect>
        </g>
        <linearGradient spreadmethod="pad"
                        id="gradient-linear-gauge-series-simple-bar-undefined-0"
                        y1="0%"
                        y2="0%"
                        x1="0%"
                        x2="100%"
        >
          <stop offset="0%"
                stop-opacity="0.3"
                stop-color="#2d60e8"
          >
          </stop>
          <stop offset="80%"
                stop-opacity="1"
                stop-color="#2d60e8"
          >
          </stop>
        </linearGradient>
      </g>
    </svg>
  </div>
</div>
`;
